packages/discovery-react-components/src/components/DocumentPreview/components/PdfViewerHighlight/utils/textLayout/dom.ts

```diff
@@ -0,0 +1,67 @@
+import { getTextNodeAndOffset, uniqRects } from 'utils/document/documentUtils';
+import { Bbox, TextSpan } from '../../types';
+import { BOTTOM, LEFT, RIGHT, TOP } from '../common/bboxUtils';
+import { END, START } from '../common/textSpanUtils';
+import { TextLayoutCell } from './types';
+
+const debugOut = require('debug')?.('pdf:textLayout:dom');
+function debug(...args: any) {
+  debugOut?.apply(null, args);
+}
+
+export function getAdjustedCellByOffsetByDom(
+  cell: TextLayoutCell,
+  textSpan: TextSpan,
+  spanElement: HTMLElement,
+  scale: number
+): Bbox | null {
+  if (!(spanElement.firstChild instanceof Text) || !(spanElement.lastChild instanceof Text)) {
+    debug('unexpected. span dont have text node');
+    return null;
+  }
+
+  const beginOffset = textSpan[START];
+  const endOffset = Math.min(cell.text.length, textSpan[END]);
+
+  let left = cell.bbox[LEFT];
+  let right = cell.bbox[RIGHT];
+  const top = cell.bbox[TOP];
+  const bottom = cell.bbox[BOTTOM];
+
+  // convert offset
+  function getAdjustedOffset(orgOffset: number) {
+    return orgOffset;
+  }
+  try {
+    const { textNode: beginTextNode, textOffset: beginTextOffset } =
+      beginOffset > 0
+        ? getTextNodeAndOffset(spanElement, getAdjustedOffset(beginOffset))
+        : { textNode: spanElement.firstChild, textOffset: 0 };
+    const { textNode: endTextNode, textOffset: endTextOffset } =
+      endOffset > 0
+        ? getTextNodeAndOffset(spanElement, getAdjustedOffset(endOffset))
+        : { textNode: spanElement.lastChild, textOffset: spanElement.lastChild.length };
+
+    debug('finding text node for: ', cell.text);
+    debug('  textContent: ', beginTextNode.textContent);
+    debug('  beginOffset: ', beginTextOffset);
+    debug('  textContent: ', endTextNode.textContent);
+    debug('    endOffset: ', endTextOffset);
+
+    const range = document.createRange();
```

Comment:
1. Does `range` need to be deleted/destroyed after it's done being used here? Or will the garbage collector take care of it?
2. This code has some big similarities to `createFieldRects` in `documentUtils.ts`. Should the be refactored to make use of same underlying code? Or would that be too complex?

Reply:
> 1. Does range need to be deleted/destroyed after it's done being used here?

From [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Range/detach), we used to use `Range.detach()`, but it does nothing in recent browsers.

Reply:
> 2. we can extract a common logic of iterating over rectangles on a range.